src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -21,80 +21,81 @@ namespace NuGet.CommandLine.XPlat
-        private const string ProjectName = "MSBuildProjectName";
+        private const int GenericSuccessExitCode = 0;
+        private const int GenericFailureExitCode = 1;
         private Dictionary<PackageSource, SourceRepository> _sourceRepositoryCache;
 
         public ListPackageCommandRunner()
         {
             _sourceRepositoryCache = new Dictionary<PackageSource, SourceRepository>();
         }
 
-        public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
+        public async Task<int> ExecuteCommandAsync(ListPackageArgs listPackageArgs)
         {
+            IReportRenderer reportRenderer = listPackageArgs.Renderer;
+            (int exitCode, ListPackageReportModel reportModel) = await GetReportDataAsync(listPackageArgs);
+            reportRenderer.AddProjectReport(reportModel);
+            return exitCode;
+        }
+
+        internal async Task<(int, ListPackageReportModel)> GetReportDataAsync(ListPackageArgs listPackageArgs)
+        {
+            // It's important not to print anything to console from below methods and sub method calls, because it'll affect both json/console outputs.
+            var listPackageReportModel = new ListPackageReportModel(listPackageArgs);
             if (!File.Exists(listPackageArgs.Path))
             {
-                Console.Error.WriteLine(string.Format(CultureInfo.CurrentCulture,
+                listPackageArgs.Renderer.AddProblem(errorText: string.Format(CultureInfo.CurrentCulture,
                         Strings.ListPkg_ErrorFileNotFound,
-                        listPackageArgs.Path));
-                return;
+                        listPackageArgs.Path),
+                        problemType: ProblemType.Error);
+                return (GenericFailureExitCode, listPackageReportModel);
             }
             //If the given file is a solution, get the list of projects
             //If not, then it's a project, which is put in a list
             var projectsPaths = Path.GetExtension(listPackageArgs.Path).Equals(".sln") ?
                            MSBuildAPIUtility.GetProjectsFromSolution(listPackageArgs.Path).Where(f => File.Exists(f)) :
                            new List<string>(new string[] { listPackageArgs.Path });
 
-            var autoReferenceFound = false;
-            var msBuild = new MSBuildAPIUtility(listPackageArgs.Logger);
+            MSBuildAPIUtility msBuild = listPackageReportModel.MSBuildAPIUtility;
 
-            //Print sources, but not for generic list (which is offline)
-            if (listPackageArgs.ReportType != ReportType.Default)
-            {
-                Console.WriteLine();
-                Console.WriteLine(Strings.ListPkg_SourcesUsedDescription);
-                ProjectPackagesPrintUtility.PrintSources(listPackageArgs.PackageSources);
-                Console.WriteLine();
-            }
-
-            foreach (var projectPath in projectsPaths)
+            foreach (string projectPath in projectsPaths)
```

Comment:
Can we extract everything inside this `foreach` loop into a method, so we separate "get the list of all projects and iterate it" in a different method to "get the data model for a single proejct"?  I think it will make the code easier to read (and test).

Reply:
I'd resolve simple threads. It makes reviewing a lot easier imo. 

Yes, I know I can disable comments, but I want to be able to view my own comments. Otherwise I just end up with 3 different views :D